plutus-ir/src/Language/PlutusIR/Compiler/Let.hs

```diff
@@ -62,45 +62,47 @@ compileLet :: Compiling m e uni a => LetKind -> PIRTerm uni a -> DefT SharedName
 compileLet kind = \case
     Let p r bs body -> withEnclosing (const $ LetBinding r p) $ case r of
             -- See Note [Right-associative compilation of let-bindings for linear scoping]
-            NonRec -> lift $ foldM (compileNonRecBinding kind) body (reverse bs)
+            NonRec -> lift $ foldM (compileNonRecBinding kind) body (NE.reverse bs)
             Rec    -> compileRecBindings kind body bs
     x -> pure x
 
 compileRecBindings
     :: Compiling m e uni a
     => LetKind
     -> PIRTerm uni a
-    -> [Binding TyName Name uni (Provenance a)]
+    -> NE.NonEmpty (Binding TyName Name uni (Provenance a))
     -> DefT SharedName uni (Provenance a) m (PIRTerm uni a)
-compileRecBindings kind body bs
-    | null typeBinds = compileRecTermBindings kind body termBinds
-    | null termBinds = lift $ compileRecTypeBindings kind body typeBinds
-    | otherwise      = lift $ getEnclosing >>= \p -> throwing _Error $ CompilationError p "Mixed term and type bindings in recursive let"
+compileRecBindings kind body bs =
+  if null mTailGroups
+  -- only one single group should appear, we do not allow mixing of bind styles
+  then case NE.head groupSingle of
+         TermBind {} -> compileRecTermBindings kind body groupSingle
+         DatatypeBind {} ->  lift $ compileRecDataBindings kind body groupSingle
+         TypeBind {} -> lift $ getEnclosing >>= \p -> throwing _Error $ CompilationError p "Type bindings cannot appear in recursive let, use datatypebind instead"
+  else lift $ getEnclosing >>= \p -> throwing _Error $ CompilationError p "Mixed term/type/data bindings in recursive let"
     where
-        (termBinds, typeBinds) = partition (\case { TermBind {} -> True ; _ -> False; }) bs
+        groupSingle :| mTailGroups  = NE.groupWith1 (\case { TermBind {} -> LT ; TypeBind {} -> EQ; _ -> GT }) bs
```

Comment:
This is a bit weird, I think you're using `Ordering` as a random type to group by which happens to have three elements? I think it would actually be clearer to just use `1,2,3`, since for a while I thought the ordering output had some meaning.

Also, this seems suspect, since `group` and `groupWith` group *preserving the ordering*. `groupAllWith1` does the right thing. In fact this doesn't matter, since you only care about the case where there's precisely one group, but that's pretty subtle and could either do with a less subtle implementation or a careful comment.

Reply:
Both your assumptions are correct. I will use 1,2,3 instead and add a comment of why we don't have to use `groupAll` in this case.

Reply:
It is a bit difficult to explain why `groupAllWith1` is an overkill here, but I tried my best.